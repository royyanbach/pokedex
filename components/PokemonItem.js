import React, { Component, createRef } from "react";

export default class PokemonItem extends Component {
  constructor(props) {
    super(props);

    const pokemonId = this.props.pokemonDetail.url.replace(/\/$/, '').split('/').pop();
    const backgroundUrl = `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/${pokemonId}.png`;

    this.pokemonId = pokemonId;
    this.el = createRef();
    this.style = {
      backgroundImage: `url(${backgroundUrl})`
    }
  }

  componentDidMount() {
    if (this.props.observer) {
      this.props.observer.observe(this.el.current);
    }
  }

  componentDidUpdate() {
    if (this.props.observer) {
      this.props.observer.observe(this.el.current);
    }
  }

  render() {
    return(
      <li className="pokemon-item-wrapper" style={ this.style } onClick={ () => this.props.onSelectPokemon(this.pokemonId) } ref={ this.el }>
        { this.props.pokemonDetail.name }
      </li>
    );
  }
}
