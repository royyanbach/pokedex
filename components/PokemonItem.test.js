import React from 'react';
import { render, unmountComponentAtNode } from "react-dom";
import renderer from 'react-test-renderer';
import { act } from "react-dom/test-utils";

import PokemonItem from './PokemonItem';

let container = null;
beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it('Render list with correctly', () => {
  const pokemonPreviewData = {
    name: 'bulbasaur',
    url: 'https://pokeapi.co/api/v2/pokemon/1/'
  };

  const component = renderer.create(
    <PokemonItem pokemonPreviewData={ pokemonPreviewData }/>,
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('Handle click event', () => {
  const onSelectPokemon = jest.fn();
  const id = 1;
  const pokemonPreviewData = {
    name: 'bulbasaur',
    url: `https://pokeapi.co/api/v2/pokemon/${id}/`
  };

  act(() => {
    render(<PokemonItem pokemonPreviewData={ pokemonPreviewData } onSelectPokemon={ onSelectPokemon } />, container);
  });

  const renderedItem = document.querySelector('.pokemon-item-wrapper');

  act(() => {
    renderedItem.dispatchEvent(new MouseEvent('click', { bubbles: true }));
  });

  expect(onSelectPokemon).toHaveBeenCalledWith(id);
});

it('Observe it owns element if observer is given', () => {
  const observer = {
    observe: jest.fn(),
  };

  const pokemonPreviewData = {
    name: 'bulbasaur',
    url: 'https://pokeapi.co/api/v2/pokemon/1/'
  };
  
  act(() => {
    render(<PokemonItem pokemonPreviewData={ pokemonPreviewData } observer={ observer }/>, container);
  });

  expect(observer.observe).toHaveBeenCalled();
})
