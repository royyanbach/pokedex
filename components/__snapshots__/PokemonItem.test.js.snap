// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render list with correctly 1`] = `
<li
  className="pokemon-item-wrapper"
  onClick={[Function]}
  style={
    Object {
      "backgroundImage": "url(https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png)",
    }
  }
>
  bulbasaur
</li>
`;
